### PR TITLE
YD-425 Revived AnalysisEngineServiceTests

### PR DESCRIPTION
**Pull request by Bert-R**

### CODE REVIEW

File: analysisservice/src/main/java/nu/yona/server/analysis/service/AnalysisEngineService.java [hunk 1]
```diff
@@ -359,9 +359,9 @@ private DayActivity createNewDayActivity(UserAnonymized userAnonymizedEntity, Ac
 		{
 			weekActivity = WeekActivity.createInstance(userAnonymizedEntity, matchingGoal, startOfWeek.getZone(),
 					startOfWeek.toLocalDate());
+			matchingGoal.addWeekActivity(weekActivity);
 		}
 		weekActivity.addDayActivity(dayActivity);
-		matchingGoal.addWeekActivity(weekActivity);
```
Comment (Bert-R) [inline analysisservice/src/main/java/nu/yona/server/analysis/service/AnalysisEngineService.java]: Found this little bug through the unit test. In practice, it currently doesn't do any harm, as the foreign key relationship causes the entity to be present just once.

File: analysisservice/src/test/java/nu/yona/server/analysis/service/AnalysisEngineServiceTests.java [hunk 2]
```diff
@@ -376,15 +396,16 @@ public void messageAggregation()
 		// Verify that the cache is used to check existing activity
 		verify(mockAnalysisEngineCacheService, times(3)).fetchLastActivityForUser(userAnonId, gamblingGoal.getId());
 		// Verify that there is no new conflict message sent.
-		verify(mockMessageService, never()).sendMessageAndFlushToDatabase(any(), eq(anonMessageDestination));
+		ArgumentCaptor<MessageDestination> messageDestination = ArgumentCaptor.forClass(MessageDestination.class);
```
Comment (aristotelos) [inline analysisservice/src/test/java/nu/yona/server/analysis/service/AnalysisEngineServiceTests.java]: Can you extract a method `verifyNoGoalConflictMessageSent` too?

File: analysisservice/src/test/java/nu/yona/server/analysis/service/AnalysisEngineServiceTests.java [hunk 3]
```diff
@@ -244,15 +275,31 @@ public void conflictInterval()
...
+	@Test
+	public void noMessageSentWhenSecondConflictWithinConflictInterval()
+	{
+		mockExistingActivity(gamblingGoal, nowInAmsterdam());
+
+		Set<String> conflictCategories = new HashSet<>(Arrays.asList("lotto"));
+		service.analyze(userAnonId, new NetworkActivityDto(conflictCategories, "http://localhost/test1", Optional.empty()));
+
+		// Verify that there is a new conflict message sent.
+		ArgumentCaptor<MessageDestination> messageDestination = ArgumentCaptor.forClass(MessageDestination.class);
```
Comment (aristotelos) [inline analysisservice/src/test/java/nu/yona/server/analysis/service/AnalysisEngineServiceTests.java]: Can you extract a method `verifyNoGoalConflictMessageSent` too?

File: analysisservice/src/test/java/nu/yona/server/analysis/service/AnalysisEngineServiceTests.java [hunk 4]
```diff
@@ -460,58 +463,103 @@ public void activityOnNewDay()
...
 		// Verify that there is a new conflict message sent.
-		verify(mockMessageService, times(1)).sendMessageAndFlushToDatabase(any(), eq(anonMessageDestination));
+		verifyGoalConflictMessageSent(gamblingGoal);
+
 		// Verify that a database lookup was done finding the existing DayActivity to update
 		verify(mockDayActivityRepository, times(1)).findOne(userAnonId, now.toLocalDate(), gamblingGoal.getId());
+
 		// Verify that the day was updated
-		verify(mockDayActivityRepository, times(1)).save(dayActivity);
+		ArgumentCaptor<UserAnonymized> userAnonymizedCaptor = ArgumentCaptor.forClass(UserAnonymized.class);
```
Comment (aristotelos) [inline analysisservice/src/test/java/nu/yona/server/analysis/service/AnalysisEngineServiceTests.java]: A method could be extracted for this day activity update pattern too. Then we can remove the comments :).

File: analysisservice/src/test/java/nu/yona/server/analysis/service/AnalysisEngineServiceTests.java (hunk 2 above)
Comment (aristotelos) [inline analysisservice/src/test/java/nu/yona/server/analysis/service/AnalysisEngineServiceTests.java]: Ehm, it already exists, `verifyNoMessagesCreated`.

File: analysisservice/src/test/java/nu/yona/server/analysis/service/AnalysisEngineServiceTests.java (hunk 3 above)
Comment (aristotelos) [inline analysisservice/src/test/java/nu/yona/server/analysis/service/AnalysisEngineServiceTests.java]: Ehm, it already exists, `verifyNoMessagesCreated`.

File: analysisservice/src/test/java/nu/yona/server/analysis/service/AnalysisEngineServiceTests.java [hunk 5]
```diff
@@ -215,10 +246,10 @@ public void getRelevantSmoothwallCategories()
 	}
 
 	/*
-	 * Tests that two conflict messages are generated when the conflict interval is passed.
+	 * Tests that a messages is sent when another conflict occurs after the interval
```
Comment (Bert-R) [inline analysisservice/src/test/java/nu/yona/server/analysis/service/AnalysisEngineServiceTests.java]: Done

File: analysisservice/src/test/java/nu/yona/server/analysis/service/AnalysisEngineServiceTests.java [hunk 6]
```diff
@@ -244,15 +275,31 @@ public void conflictInterval()
 		service.analyze(userAnonId, new NetworkActivityDto(conflictCategories, "http://localhost/test1", Optional.empty()));
 
 		// Verify that there is a new conflict message sent.
-		ArgumentCaptor<MessageDestinationDto> messageDestination = ArgumentCaptor.forClass(MessageDestinationDto.class);
```
Comment (Bert-R) [inline analysisservice/src/test/java/nu/yona/server/analysis/service/AnalysisEngineServiceTests.java]: Yes, sorry. At the time I refactored this test case, I didn't see the helper methods yet. After I saw them, I didn't go back. Corrected it now.

File: analysisservice/src/test/java/nu/yona/server/analysis/service/AnalysisEngineServiceTests.java (hunk 3 above)
Comment (Bert-R) [inline analysisservice/src/test/java/nu/yona/server/analysis/service/AnalysisEngineServiceTests.java]: Done.

File: analysisservice/src/test/java/nu/yona/server/analysis/service/AnalysisEngineServiceTests.java (hunk 2 above)
Comment (Bert-R) [inline analysisservice/src/test/java/nu/yona/server/analysis/service/AnalysisEngineServiceTests.java]: Done.

File: analysisservice/src/test/java/nu/yona/server/analysis/service/AnalysisEngineServiceTests.java [hunk 7]
```diff
@@ -264,30 +311,29 @@ public void messageCreatedOnMatch()
...
-		// Verify that there is a day activity update.
-		// first the DayActivity is created and then updated with the Activity
-		// so (create, update) = 2 times save
-		ArgumentCaptor<DayActivity> dayActivity = ArgumentCaptor.forClass(DayActivity.class);
-		verify(mockDayActivityRepository, times(2)).save(dayActivity.capture());
-		Activity activity = dayActivity.getValue().getLastActivity();
-		assertThat("Expect right user set to activity", dayActivity.getValue().getUserAnonymized(), equalTo(userAnonEntity));
+		// Verify there is a an activity in a day activity inside a week activity
+		ArgumentCaptor<UserAnonymized> userAnonymizedCaptor = ArgumentCaptor.forClass(UserAnonymized.class);
+		verify(mockUserAnonymizedService, atLeastOnce()).updateUserAnonymized(userAnonymizedCaptor.capture());
```
Comment (Bert-R) [inline analysisservice/src/test/java/nu/yona/server/analysis/service/AnalysisEngineServiceTests.java]: I don't understand what you have in mind.
The code here verifies that ``updateUserAnonymized`` is called at least once and that it is called for the object that is equal to ``useranonEntity``.
What can be removed while still asserting the same things?

File: analysisservice/src/test/java/nu/yona/server/analysis/service/AnalysisEngineServiceTests.java (hunk 4 above)
Comment (Bert-R) [inline analysisservice/src/test/java/nu/yona/server/analysis/service/AnalysisEngineServiceTests.java]: Though it makes sense, I leave that for a next improvement round. The test works again and is considerably better maintainable. We need to leave some work for later times :)

**Comment by Bert-R** [comment]
Pushed an update for most review comments.
If you're OK with it, you can merge it. The build server issues are resolved.

File: analysisservice/src/test/java/nu/yona/server/analysis/service/AnalysisEngineServiceTests.java (hunk 7 above)
Comment (aristotelos) [inline analysisservice/src/test/java/nu/yona/server/analysis/service/AnalysisEngineServiceTests.java]: You can write it as `verify(mockUserAnonymizedService, atLeastOnce()).updateUserAnonymized(userAnonEntity);`, isn't it?

File: analysisservice/src/test/java/nu/yona/server/analysis/service/AnalysisEngineServiceTests.java (hunk 4 above)
Comment (aristotelos) [inline analysisservice/src/test/java/nu/yona/server/analysis/service/AnalysisEngineServiceTests.java]: OK

File: analysisservice/src/test/java/nu/yona/server/analysis/service/AnalysisEngineServiceTests.java (hunk 7 above)
Comment (Bert-R) [inline analysisservice/src/test/java/nu/yona/server/analysis/service/AnalysisEngineServiceTests.java]: You're totally right.

**Pull request merged**